src/consumers/manager.rs

```diff
@@ -88,22 +99,85 @@ impl<C> ConsumerManager<C> {
     /// Returns a new consumer manager set up to use the given permit pool. This meant to use a
     /// shared pool of permits with other consumer managers to manage the amount of simultaneous
     /// work, so the Semaphore must be wrapped in an [`Arc`].
-    pub fn new(permit_pool: Arc<Semaphore>, stream: NatsStream) -> ConsumerManager<C> {
-        ConsumerManager {
+    ///
+    /// This function will attempt to populate this itself with all existing consumers. Any errors
+    /// that occur during population will only log and not error out as it is recoverable. Because
+    /// of this, it requires something that can generate the desired worker
+    pub async fn new<W, F>(
+        permit_pool: Arc<Semaphore>,
+        stream: NatsStream,
+        worker_generator: F,
+    ) -> ConsumerManager<C>
+    where
+        W: Worker + Send + Sync + 'static,
+        C: Stream<Item = Result<ScopedMessage<W::Message>, async_nats::Error>>
+            + CreateConsumer<Output = C>
+            + Send
+            + Unpin
+            + 'static,
+        F: WorkerCreator<Output = W>,
+    {
+        let mut manager = ConsumerManager {
             handles: Arc::new(RwLock::new(HashMap::default())),
             permits: permit_pool,
             stream,
             phantom: PhantomData,
-        }
+        };
+
+        let handles: HashMap<String, JoinHandle<WorkResult<()>>> = manager.stream.consumers().filter_map(|res| async {
+            let info = match res {
+                Ok(info) => info,
+                Err(e) => {
+                    error!(error = %e, "Error when trying to read current consumers");
+                    return None
+                }
+            };
+            // TODO: This is somewhat brittle as we could change naming schemes, but it is
+            // good enough for now. We are just taking the name (which should be of the
+            // format `<consumer_prefix>_<lattice_id>`), but this makes sure we are always
+            // getting the last thing in case of other underscores
+            let lattice_id = match info.name.split('_').last() {
+                Some(id) => id,
+                None => return None,
+            };
+            // NOTE(thomastaylor312): It might be nicer for logs if we add an extra param for a
+            // friendly consumer manager name
+            trace!(%lattice_id, "Adding consumer for lattice");
+
+            let worker = match worker_generator.create(lattice_id).await {
+                Ok(w) => w,
+                Err(e) => {
+                    error!(error = %e, %lattice_id, "Unable to add consumer for lattice. Error when generating worker");
+                    return None;
+                }
+            };
+
+            // NOTE: If this becomes a bottle neck at scale, we could rewrite to take a single write lock and then 
```

Comment:
And then.... And then what?? 😉

Reply:
And then...I changed my mind and did what I was writing anyway 🤣